crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+}
+
+int ossl_ml_kem_parse_public_key(const uint8_t *in,
+                                 scalar *m, scalar *t,
+                                 uint8_t rho[ML_KEM_RANDOM_BYTES],
+                                 uint8_t pkhash[ML_KEM_PKHASH_BYTES],
+                                 vinfo_t vinfo,
+                                 mctx *ctx)
+{
+    /* Decode and check |t| */
+    if (!vector_decode_12(t, in, vinfo->rank))
+        return 0;
+    /* Save the matrix |m| recovery seed |rho| */
+    memcpy(rho, in + vinfo->vector_bytes, ML_KEM_RANDOM_BYTES);
+    /*
+     * Pre-compute the public key hash, needed for both encap and decap.
+     * Also pre-compute the matrix expansion, stored with the public key.
+     */
+    return hash_h(pkhash, in, vinfo->pubkey_bytes, ctx)
+        && matrix_expand(m, rho, vinfo->rank, ctx);
+}
+
+void ossl_ml_kem_encode_private_key(uint8_t *out,
+                                    const scalar *s,
+                                    const scalar *t,
+                                    const uint8_t rho[ML_KEM_RANDOM_BYTES],
+                                    const uint8_t pkhash[ML_KEM_PKHASH_BYTES],
+                                    const uint8_t z[ML_KEM_RANDOM_BYTES],
+                                    vinfo_t vinfo)
+{
+    vector_encode_12(out, s, vinfo->rank);
+    out += vinfo->vector_bytes;
+    ossl_ml_kem_encode_public_key(out, t, rho, vinfo);
+    out += vinfo->pubkey_bytes;
+    memcpy(out, pkhash, ML_KEM_PKHASH_BYTES);
+    out += ML_KEM_PKHASH_BYTES;
+    memcpy(out, z, ML_KEM_RANDOM_BYTES);
+}
+
+/* Loading of explicit private keys is a test-only interface. */
+int ossl_ml_kem_parse_private_key(const uint8_t *in,
+                                  scalar *m,
+                                  scalar *s,
+                                  scalar *t,
+                                  uint8_t rho[ML_KEM_RANDOM_BYTES],
+                                  uint8_t pkhash[ML_KEM_PKHASH_BYTES],
+                                  uint8_t z[ML_KEM_RANDOM_BYTES],
+                                  vinfo_t vinfo,
+                                  mctx *ctx)
+{
+    int rank = vinfo->rank;
+
+    /* Decode and check |s|. */
+    if (!vector_decode_12(s, in, rank))
+        return 0;
+    in += vinfo->vector_bytes;
+
+    if (!ossl_ml_kem_parse_public_key(in, m, t, rho, pkhash, vinfo, ctx))
+        return 0;
+    in += vinfo->pubkey_bytes;
+
+    /* Check public key hash. */
+    if (memcmp(pkhash, in, ML_KEM_PKHASH_BYTES) != 0)
+        return 0;
+    in += ML_KEM_PKHASH_BYTES;
+
+    memcpy(z, in, ML_KEM_RANDOM_BYTES);
+    return 1;
+}
+
+/*
+ * Key generation consumes a 32-byte RNG output plus 1 byte for the rank
+ * (domain separation) which are hashed together to produce a pair of
+ * 32-byte seeds public "rho" and private "sigma".
+ */
+int ossl_ml_kem_genkey(const uint8_t *seed,
+                       uint8_t *pubenc,
+                       scalar *m,
```

Comment:
I think it would be a good idea to state the bounds of all of the inputs of the 'scalars'.

Reply:
The sizes of the encoded public key, ciphertext and private keys vary across the three variants, so not possible to have as explicit array dimensions in the variant-neutral low-level API, except as perhaps comments.
The non-static functions check all applicable bounds, returning an error when the provided lengths are wrong, but with decapsulation, nevertheless returning a random shared key when the ciphertext length is wrong.